### PR TITLE
Crystal v0.5.1 (new formula)

### PR DESCRIPTION
**Pull request by ctate**
Open source code generator for every language, library and framework.

### CODE REVIEW

**Comment by DomT4** [comment]
Do it this way around:

```ruby
  def install
    ENV.prepend_path "PATH", "#{Formula["node"].opt_libexec}/npm/bin"
    system "npm", "install"
    libexec.install Dir["*"]
    bin.install_symlink Dir["#{libexec}/bin/*"]
  end
```

**Comment by ctate** [comment]
Done - thank you!

File: Library/Formula/crystal.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      imports:
+        readme: ReadmeGenerator
+      outputs:
+        - generator: readme
+          spec:
+            name: $name
+            version: $version
+            description: $description
+    EOS
+    cd testpath
```
Comment (DomT4) [inline Library/Formula/crystal.rb]: You'll already be in the `testpath` at this point. No need to `cd` into it.

File: Library/Formula/crystal.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
+class Crystal < Formula
+  desc "Code generator for every language, library and framework"
+  homepage "http://crystal.sh"
+  url "https://github.com/crystal/crystal/archive/v0.5.1.tar.gz"
+  sha256 "6c022ad59a10cec58856dba858a030b745333909108c08bd326c20ac0cb144fb"
+
+  depends_on "node"
+
+  def install
+    ENV.prepend_path "PATH", "#{Formula["node"].opt_libexec}/npm/bin"
```
Comment (DomT4) [inline Library/Formula/crystal.rb]: Need an `ENV["HOME"] = ".brew_home"` under this line as well.

File: Library/Formula/crystal.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    system "npm", "install"
+    libexec.install Dir["*"]
+    bin.install_symlink Dir["#{libexec}/bin/*"]
+  end
+
+  test do
+    mkdir (testpath/".crystal")
```
Comment (DomT4) [inline Library/Formula/crystal.rb]: This'll be created by the `write` below, so you're fine to scrap it.

**Comment by ctate** [comment]
Pushed all of these improvements. Thank you!

**Comment by DomT4** [comment]
Do we need to enhance the test at all? It's passing our test, but failing internal tests apparently?

```
==> /usr/local/Cellar/crystal/0.5.1/bin/crystal update
ERROR: No schema specified in call to validate.
Failed.
==> /usr/local/Cellar/crystal/0.5.1/bin/crystal build
Building from: /private/tmp/crystal20150609-4168-y9j7vh
Loading Configuration...
ERROR: No schema specified in call to validate.
Failed.
==> /usr/local/Cellar/crystal/0.5.1/bin/crystal test
```

**Comment by ctate** [comment]
Thanks for pointing that out. Fixed test and added missing folder.

File: Library/Formula/crystal.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  desc "Code generator for every language, library and framework"
+  homepage "http://crystal.sh"
+  url "https://github.com/crystal/crystal/archive/v0.5.1.tar.gz"
+  sha256 "6c022ad59a10cec58856dba858a030b745333909108c08bd326c20ac0cb144fb"
+
+  depends_on "node"
+
+  def install
+    ENV.prepend_path "PATH", "#{Formula["node"].opt_libexec}/npm/bin"
+    ENV["HOME"] = ".brew_home"
```
Comment (mikemcquaid) [inline Library/Formula/crystal.rb]: `ENV["HOME"] = "#{buildpath}/.brew_home"`, sorry.

File: Library/Formula/crystal.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      name: test
+      version: 0.1.0
+      description: test description
+      author:
+        name: Test User
+        email: test@example.com
+        url: http://example.com
+      copyright: 2015 Test
+    EOS
+    system bin/"crystal", "build"
```
Comment (mikemcquaid) [inline Library/Formula/crystal.rb]: Nice test :

**Comment by mikemcquaid** [comment]
Sorry, one final tweak.

**Comment by ctate** [comment]
Thanks! Rebased, squashed and pushed.

**Comment by HangingClowns** [comment]
Already out of date. Have 0.7.3 out, can update? http://crystal-lang.org/2015/06/07/crystal-0.7.3-released.html

**Comment by ctate** [comment]
@HangingClowns That's the wrong project. Please note the latest version of Crystal is v0.5.1: https://github.com/crystal/crystal/releases/tag/v0.5.1

Hope that helps to clarify. Thanks!

**Comment by HangingClowns** [comment]
So embarassing... seems projects are running out of words :smile:

**Comment by mikemcquaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

For future reference the preferred commit message format for simple version updates is `crystal 0.5.1 (new formula)`.

**Comment by jhass** [comment]
* https://github.com/crystal/crystal 25 stars
* https://github.com/manastech/crystal over 2,200 stars

Wouldn't it be better to name it crystal-generator or something?

**Comment by mikemcquaid** [comment]
Too late, unfortunately.

**Comment by jhass** [comment]
It's not like manastech/crystal wasn't mentioned before the merge, might be something to pay attention to in the future.

**Comment by waj** [comment]
I don't think Crystal language would have been accepted because it installs binaries (Crystal is written in Crystal). That's why I never sent a PR. The downside now is we have to install with `brew install manastech/crystal/crystal` to avoid the name clashing.

**Comment by xu-cheng** [comment]
> because it installs binaries (Crystal is written in Crystal).

We can handle this kind of bootstrap issue. You can see `mono.rb` for example. You are welcome to send PR for Crystal language.

> The downside now is we have to install with brew install manastech/crystal/crystal to avoid the name clashing.

Using tap is not recommended either, as in that case users won't be able to install this PR's formula and manastech/crystal at the same time. The best solution would be naming manastech/crystal as `crystal-long`

**Comment by havenwood** [comment]
It seems an awful lot of people are going to `brew install crystal` expecting the Crystal language and be confused why they're getting a coffeescript script whose documentation says to install it with npm.

**Comment by will** [comment]
`crystal` should be the language not a thing you can also install with npm

**Comment by asterite** [comment]
I would at least expect the tool to become a bit more popular before taking a homebrew formula name of something that's at least more popular (just by counting the starts in the repo, the [IRC activity](http://irclog.whitequark.org/crystal-lang) and all the [talks](https://github.com/nob-suz/crystal/wiki/2.-Slides-or-presentation-of-Crystal-language) that are happening worldwide) :-( 

I personally didn't submit this formula for our project because it's not very popular yet, not even at version 1.0.

**Comment by xu-cheng** [comment]
Sorry but for technical reason, we don't support formula rename. You can see the historical issue in #14374. I will lock this thread now. And feel free to submit PR for crystal language.

**Pull request closed without merging**